### ISSUE TITLE
Data from image is not copied to tmpfs named volumes upon container creation

### ISSUE THREAD

**Issue by dmandalidis** [docker/docker#27068]
**Description**

Data is not copied to tmpfs named volumes upon container creation.

**Steps to reproduce the issue:**
1. FROM busybox:latest RUN mkdir /data && touch /data/atestfile CMD ls -la /data
2. docker build -t foo-image .
3. docker create -v foo1:/data foo-image # /data/atestfile exists in foo1
4. docker volume create --driver local --opt type=tmpfs --opt device=tmpfs --opt o=size=1m,uid=0 --name foo2 # repeat step 3 -> /data/atestfile doesn't exist in foo2
5. docker volume create --name foo3 -o o=addr=10.0.0.1,rw,soft -o device=:/share -o type=nfs # repeat step 3 -> /data/atestfile exists in foo3

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:57:25 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:57:25 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 5
 Running: 1
 Paused: 0
 Stopped: 4
Images: 783
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 603
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay null host bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 3.19.0-66-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 993.2 MiB
Name: dock-02
ID: DBYH:VD7W:KJRG:5G2R:E62M:HLXE:4ZG6:GSCI:JIJN:YLXP:6HI2:63O5
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: http://proxyblue.trasys.be:80
Https Proxy: http://proxyblue.trasys.be:80
No Proxy: 172.17.52.20
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: true
Insecure Registries:
 172.17.52.20:9090
 127.0.0.0/8
```

**Comment by dmandalidis** [comment]
debug output at http://pastebin.com/KPiPpefk

**Comment by cpuguy83** [comment]
Thanks for opening a new issue.

We should look at this since there may be a bug, but also I'm not sure this is a particularly valid use case.
Every time the container is started it will get a new tmpfs mounted to it, meanwhile volume initialization only happens when a container is created.

**Comment by dmandalidis** [comment]
Indeed, it doesn't make so much sense to initialize a tmpfs on create. I tested this on startup and it still doesn't happen (http://pastebin.com/P7m1fZ0e) but there's an error log stmt now :

`level=error msg="containerd: notify OOM events" error="cgroup path for memory not found"`

is this relevant?

**Comment by justincormack** [comment]
@dmandalidis no the cgroup path error is unrelated (this is only in some kernel versions and configurations).

**Comment by mterron** [comment]
This makes tmpfs useless for containers that do not run as root. The pattern OP posted is how it should work, in particular for /run.

**Comment by bernardjech** [comment]
I actually wish tmpfs initializing on create was possible. I have an image of PostgreSQL database that I use for unit testing. Each unit test needs the database clean, and the database content is discarded at the end of the test. Each test creates a new database container from the image, the tests run, and the container is removed at the end. I wanted to use tmpfs for the pgdata directory in order to speed up the tests, but alas, the tmpfs is not initialized from the image.

**Comment by thaJeztah** [comment]
Had a quick look why the content is not "copied", using a custom volume that's using `tmpfs`;

```bash
docker volume create --driver local \
    --opt type=tmpfs \
    --opt device=tmpfs \
    --opt o=size=100m,uid=1000 \
    mytmpfsvolume
```
Starting a container that mounts that volume, the daemon logs will show;

```
DEBU[2020-08-17T11:35:24.509005957Z] copying image data from 519e2eab1aac240d62899cb439c7fb88e4dce7ad277d5369935e84fb47ae720a:/foo, to mytmpfsvolume
DEBU[2020-08-17T11:35:24.539000047Z] Calling POST /v1.41/containers/519e2eab1aac240d62899cb439c7fb88e4dce7ad277d5369935e84fb47ae720a/attach?stderr=1&stdout=1&stream=1
DEBU[2020-08-17T11:35:24.539314722Z] attach: stderr: begin
DEBU[2020-08-17T11:35:24.539341909Z] attach: stdout: begin
DEBU[2020-08-17T11:35:24.540338407Z] Calling POST /v1.41/containers/519e2eab1aac240d62899cb439c7fb88e4dce7ad277d5369935e84fb47ae720a/wait?condition=removed
DEBU[2020-08-17T11:35:24.541527196Z] Calling POST /v1.41/containers/519e2eab1aac240d62899cb439c7fb88e4dce7ad277d5369935e84fb47ae720a/start
```

So, the data is copied from the image to the `mytmpfsvolume` before the container is started, then the daemon attaches to the container, and the container is started.

The `copying image data` debug log is printed here; https://github.com/moby/moby/blob/ce21a168336745a77d401762f2baf4b5cef15751/daemon/create_unix.go#L90-L92

Which is where all volumes are propagated (if needed), which goes to https://github.com/moby/moby/blob/a93e9eb0e95c6b2db723d0d2391e4559554c8597/container/container_unix.go#L125

So, while copying is actually done, a temporary mount is used for the volume; https://github.com/moby/moby/blob/a93e9eb0e95c6b2db723d0d2391e4559554c8597/container/container_unix.go#L138-L142

Then files are copied https://github.com/moby/moby/blob/a93e9eb0e95c6b2db723d0d2391e4559554c8597/container/container_unix.go#L152

But after that, the temporary mount is unmounted; https://github.com/moby/moby/blob/a93e9eb0e95c6b2db723d0d2391e4559554c8597/container/container_unix.go#L144-L148

Given that the volume is created to use a `tmpfs` mount, the data copied to the volume before starting the container will be gone once the temporary mount is unmounted. Once the container is started,
the volume will be mounted again, but no data will be there as it was previously copied to a `tmpfs` (and thus gone).
